### ISSUE TITLE
Core Solver - Solving Rotation Attributes with value of 0.0 will cause Attribute not to solve

### ISSUE THREAD

**Issue by david-cattermole** [david-cattermole/mayaMatchMoveSolver#54]
If you add a Rotation attribute into the solver and it contains a value of 0.0, the cminpack_lm solver will not change the value.

This might be caused by using the "mode=1" attribute of CMinpack lmdiff_ function.

>  mode is an integer input variable. If mode = 1, the variables will be scaled internally. If mode = 2, the scaling is specified by the input diag. Other values of mode are equivalent to mode = 1. 

A likely fix is to add an internal offset value to the attribute value, and let the solver use that value, then remove the offset value before setting back into Maya. This offset value may need to change for different attributes, rather than using a specific offset value for all attribute types (rotation verses translation for example).

**Comment by david-cattermole** [comment]
This bug cannot be reproduced. We will re-open or create a new issue when we can reproduce it.

**Comment by david-cattermole** [comment]
It has been reported that not only is this issue still happening, but the problem occurs when the cminpack solver tries to solve a parameter at exactly 0.0.

**Comment by valmikchavan** [comment]
Hey David,

I have noticed too, after projecting bundles, only translation is getting
solved, if we add any dacimal values in rotation then it's working.

Thank you!

**Comment by david-cattermole** [comment]
Yes, unfortunately, it's a known problem with the CMinpack library, but the older "levmar" didn't have the problem.

As you said, the work around is to add a small value to the rotation value, like: 0.00001, and it will solve to a non-zero value. However if the value needing to be solved must be exactly 0.0, then it also won't solve.

Another workaround is to set the value to exactly 360.0, and solve as normal. This will equal exactly the same value as 0.0, but will solve correctly.

I'm working on adding this workaround into the solver, so users don't need to do it.

David

**Comment by valmikchavan** [comment]
Hey David,

Thanks for update.

**Comment by david-cattermole** [comment]
This bug appears to be fixed in this branch.
I have added two new mmSolver command flags for each attribute; offset and scale.
Offset will add to the internal value being solved by cminpack (or levmar), likewise the scale value will be used to multiply the internal value being solved. 

During tests changing the "scale" value can cause solve instability. This may be caused by the Delta value of the solver not being large enough, or too large, to properly detect error values.

The rotation attribute problem appears to be fixed by offsetting the internal value by 360.0 for all Angle attribute types. No offsets or scales are set for any other attribute types, when using the mmSolver.api module.

More tests should be done before merging this into the master branch.

**Comment by david-cattermole** [comment]
Also note, before this ticket, the solver was actually calculating rotations in Radians. This latest commit changes that. mmSolver now calculates rotations as angles, meaning the values may be much large than before, relative to the transform values. 

With luck, we will be able to "normalise" the input attribute values using the offset and scale flags (explained above) to address issue #26.

**Comment by david-cattermole** [comment]
I have added new features into the 'lm_der' solver engine to address this issue. I am not aware of it happening anymore. This issue can be closed.
